### ISSUE TITLE
SVG foreignObject to render HTML

### ISSUE THREAD

**Issue by harrywincup** [cyclejs/core#345]
Hi,

I've been using SVG `foreignObject` to render some HTML inside of an SVG chart. I've just upgraded to the "Diversity" versions to get the event bubbling fix for `mouseenter / mouseleave`, but now the `foreignObject` content no longer renders.

Any idea if this is a bug in `@cycle/dom`, or i'm doing something wrong?

```javascript
svg.foreignObject({ attrs: { width: '100', height: '50' }}, [
    body({ attrs: { xmlns: 'http://www.w3.org/1999/xhtml' }}, [
        p('THIS IS A PARAGRAPH THAT CAN WRAP TEXT')
    ])
])
```

Any pointers much appreciated, thanks!

**Comment by laszlokorte** [comment]
Probably because of snabbdom setting the namespace of all children of an svg tag recursively: https://github.com/paldepind/snabbdom/blob/master/h.js#L29-L31

I guess it sets the NS of your body and p tag to svg as well.

**Comment by harrywincup** [comment]
@laszlokorte thanks for the tip, it looks like you're right. Do you know if there's a way to work around it? I've just tried disabling the `ns` property addition in the `addNS` function so I can set the namespace in my own code, but the results are not good.

**Comment by laszlokorte** [comment]
The guess the proper solution  would be to fix snabbdom for example by adding an additional condition to stop the `addNS` recursion if a a `foreignObject` element is hit or by providing a way to opt out of the recursing by specifying a custom (empty) NS.

Until it is fixed you could work around by doing:

```
const fo = svg.foreignObject({ attrs: { width: '100', height: '50' }}, [
])

// this must happen after the foreignObject has been embedded into the svg tag (ie after the addNS call has happened):
fo.children = [
    body({ attrs: { xmlns: 'http://www.w3.org/1999/xhtml' }}, [
        p('THIS IS A PARAGRAPH THAT CAN WRAP TEXT')
    ])
];
```

**Comment by Hypnosphi** [comment]
Can this be done with a hook?

**Comment by harrywincup** [comment]
@Hypnosphi This is my first time using Snabbdom, so i'm no expert ... However, it appears that using the `init` hook to hijack the node's `data` works:
```javascript
svg.foreignObject({ attrs: { width: '100%', height: '150' }}, [
    body({ style: { background: 'blue' }, hook: { init: (vNode) => { vNode.data.ns = 'http://www.w3.org/1999/xhtml' }}}, [
        p('THIS IS A PARAGRAPH THAT CAN WRAP TEXT')
    ])
])
```

_However_, the `p` in the example above does not render because every node needs the namespace set, which feels a bit off.

It seems to me that the underlying issue is @cycle/dom incorrectly assuming that all children of an `svg` node belong in the svg namespace (https://github.com/cyclejs/dom/blob/master/src/hyperscript.ts#L15-L26).

I tried patching `addNS` locally so it switches to the `xhtml` namespace for all child elements of a `foreignObject` node, which also works but currently feels very shoehorned.

Not really sure what the best approach is here.

**Comment by laszlokorte** [comment]
@harrywincup right, cycle comes with it's own `h` helper function instead of using the one that `snabbdom` provides. So the issues needs to be fixed inside cycle indeed. Anyway the cycle `h` function is a modified version of snabbdoms `h` so the error exist in both snabbdom and cycle.

I think stopping the `addNS` recursing if an `foreignObject` node is hit sounds like a good solution.

**Comment by harrywincup** [comment]
I'll make a PR if I can get the code going on my machine ... Currently seeing errors from `snabbdom-to-html` when I run `npm test`

**Comment by harrywincup** [comment]
https://github.com/cyclejs/dom/pull/130

**Comment by staltz** [comment]
Fixed in cycle dom v10.0.5 thanks to Harry's PR.
